services/supervisor/tests/supervisor/main.rs

```diff
@@ -63,21 +64,24 @@ fn assert_count_is_not_set(api: &TestKitApi, service_name: &'static str) {
     assert!(response.is_err());
 }
 
-fn artifact_exists(api: &TestKitApi, name: &str) -> bool {
-    let artifacts = &api.dispatcher_info().artifacts;
-    artifacts.iter().any(|a| a.name == name)
+#[allow(clippy::let_and_return)] // doesn't work otherwise
```

Comment:
Why exactly?

Reply:
Borrow checker. Here's the error messages I'm getting locally with the following code

```rust
let snapshot = testkit.snapshot();
let artifacts = snapshot.for_dispatcher().service_artifacts();
artifacts.keys().any(|artifact| artifact.name == name)
```

```text
error[E0597]: `*snapshot` does not live long enough
  --> services\supervisor\tests\supervisor\main.rs:70:21
   |
70 |     let artifacts = snapshot.for_dispatcher().service_artifacts();
   |                     ^^^^^^^^ borrowed value does not live long enough
71 |     artifacts.keys().any(|artifact| artifact.name == name)
   |     ---------------- a temporary with access to the borrow is created here ...
72 | }
   | -
   | |
   | `*snapshot` dropped here while still borrowed
   | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `exonum_merkledb::indexes::proof_map::Keys<'_, exonum::runtime::types::ArtifactId>`
   |
   = note: The temporary is part of an expression at the end of a block. Consider forcing this temporary to be dropped sooner, before the block's local variables are dropped. For example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block.

error[E0597]: `artifacts` does not live long enough
  --> services\supervisor\tests\supervisor\main.rs:71:5
   |
71 |     artifacts.keys().any(|artifact| artifact.name == name)
   |     ^^^^^^^^^-------
   |     |
   |     borrowed value does not live long enough
   |     a temporary with access to the borrow is created here ...
72 | }
   | -
   | |
   | `artifacts` dropped here while still borrowed
   | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `exonum_merkledb::indexes::proof_map::Keys<'_, exonum::runtime::types::ArtifactId>`
   |
   = note: The temporary is part of an expression at the end of a block. Consider forcing this temporary to be dropped sooner, before the block's local variables are dropped. For example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block.
```

Reply:
That's weird. In theory, `any` for iterator is not lazy, and all the variables should be dropped only after result of `any` is collected.

Here's somewhat similar working example: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c96cc80a558cf5f54812ee758a2e116d).